modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/PainlessUserTreeToIRTreePhase.java

```diff
@@ -540,6 +545,31 @@ public void injectConverter(AStatement userStatementNode, ScriptScope scriptScop
         ExpressionNode returnExpression = returnNode.getExpressionNode();
         returnNode.setExpressionNode(irInvokeCallMemberNode);
         irInvokeCallMemberNode.addArgumentNode(returnExpression);
+    }
 
+    @Override
+    public void visitCallLocal(ECallLocal userCallLocalNode, ScriptScope scriptScope) {
+        if ("$".equals(userCallLocalNode.getMethodName())) {
```

Comment:
Is there a way to arrange this so we can have `DefaultUserTreeToIRTreePhase.visitCallLocal` do this?

Reply:
After giving the data passing some more thought there isn't really an easy way to avoid specialization as part of the `PainlessUserTreeToIRTreePhase` because
* we need to unravel a single node `$` into multiple method calls
* we can't use any shortcuts directly in code like adding a `$` method to DocValuesReader or use a class binding because we explicitly rely on duck typing for this to work
* we do not want to add a base method that takes Object (`def`) as a default value because that would defeat the purpose of duck typing and we would require type method overloading as opposed to arity overloading

Moving forward refactoring our phases to rebuild the tree during each phase is the best long-term solution to this issue.